### ISSUE TITLE
[Bug][Compiler-v2] unexpected type error when using inline function with generic types

### ISSUE THREAD

**Issue by rahxephon89** [aptos-labs/aptos-core#11099]
# 🐛 Bug

When compiling the code:

```
module 0x42::M {
    use std::vector;

    struct S<Key> has copy, drop, store {
        data: vector<E<Key>>,
    }

    struct E<Key> has copy, drop, store {
        key: Key,
    }

    public fun foo<Key: store + drop>(
        map: S<Key>): vector<Key> {
        let keys: vector<Key> = vector::empty();
        let S { data } = map;
        vector::for_each(data, |e| {
            let E { key } = e;
            vector::push_back(&mut keys, key);
        });
        keys
    }

}
```
The following error message will be generated:

```
Diagnostics:
bug: unexpected type: ?10
   ┌─ tests/file-format-generator/unexpected_type.move:12:5
   │
12 │ ╭     public fun foo<Key: store + drop>(
13 │ │         map: S<Key>): vector<Key> {
14 │ │         let keys: vector<Key> = vector::empty();
15 │ │         let S { data } = map;
   · │
20 │ │         keys
21 │ │     }
   │ ╰─────^
```

I am not fully sure but it seems to be related to the inline function. cc @brianrmurphy @wrwg

**Comment by brmataptos** [comment]
Is this already fixed by https://github.com/aptos-labs/aptos-core/pull/11077 ?

**Comment by rahxephon89** [comment]
> Is this already fixed by #11077 ?

Not really. I tested this on the latest main.

**Comment by rahxephon89** [comment]
For the code:

```
module 0x42::M {
    use std::vector;
    struct E<Key> has copy, drop, store {
        key: Key,
    }
    public inline fun g<Key: store + drop>(x: E<Key>, v: |E<Key>|) {
        v(x)
    }
    public fun foo<Key: store + drop>(
        data: E<Key>) {
        let keys: vector<Key> = vector::empty();
        g(data, |e| {
            let E { key } = e;
            vector::push_back(&mut keys, key);
        });
    }
}
```

The following stackless bytecode is generated:

```
public fun M::foo<#0>($t0: M::E<#0>) {
     var $t1: vector<#0>
     var $t2: vector<#0>
     var $t3: M::E<#0>
     var $t4: M::E<#0>
     var $t5: #0
     var $t6: &mut vector<#0>
  0: $t2 := vector::empty<#0>()
  1: $t1 := infer($t2)
  2: $t3 := infer($t0)
  3: $t4 := infer($t3)
  4: $t5 := unpack M::E<?8>($t4)
  5: $t6 := borrow_local($t1)
  6: vector::push_back<#0>($t6, $t5)
  7: return ()
}
```

You can see that one type parameter `<?8>` is not instantiated, leading to the issue when generating the bytecode.
